### PR TITLE
CRM-18081 Allow search of active relationships during a custom range of dates

### PR DESCRIPTION
**Pull request by francescbassas**
* [CRM-18081: Allow search of active relationships during a custom range of dates](https://issues.civicrm.org/jira/browse/CRM-18081)

### CODE REVIEW

**Comment by jitendrapurohit** [comment]
@francescbassas Any plan to write a unit test for the same ? You may find some related examples here - tests/phpunit/api//v3/RelationshipTest.php

**Comment by francescbassas** [comment]
@jitendrapurohit sure, but I need some guidance. Which file should I to edit or create to write the unit test? Do you see the proposed changes well? I do not have much skill with English then would be nice if somebody reviewed the UI strings.

**Comment by jitendrapurohit** [comment]
As the outcome shown [on jira comment](https://issues.civicrm.org/jira/browse/CRM-18081?focusedCommentId=103604&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-103604) this PR works fine and gives the expected result. Maybe, we can optimize the calculation part to be in a single function for reports and searches. 

For unit test, you can write it under `tests//phpunit/CRM/Contact/BAO/QueryTest.php` file. Normally we could create a formValue, project it in CRM_Contact_BAO_Query and assert the whereClause() formed. Also can ping on chat.civicrm.org for any doubts.

**Comment by francescbassas** [comment]
Hi @jitendrapurohit, I optimized the calculation part to be in a a single function for reports and searchs and I renamed Active date to Active period. I tried to write the unit test, but I think it's too much for me.

**Comment by jitendrapurohit** [comment]
@francescbassas This is not working after the latest update. I even get relationship ending in 2015 when searching for relationship active in the year 2016. Did something went wrong in the optimization fix?

**Comment by francescbassas** [comment]
@jitendrapurohit I got it! The issue appears renaming active_date field to active_period. Relative dates stop working. Renaming again active_period field to active_period_date fix the issue.

**Comment by jitendrapurohit** [comment]
select2 widget broken on Advanced Search relationship tab after applying the updated patch. console is displayed as -

image

**Comment by francescbassas** [comment]
Yes, I miss the active_period to active_period_date rename on templates/CRM/Contact/Form/Search/Criteria/Relationship.tpl.

**Comment by jitendrapurohit** [comment]
message is displayed incorrectly in relationship reports if only `from` date is selected -

image

Also you can squash/fixup all your commits into single by following [interactive mode](https://git-scm.com/docs/git-rebase#_interactive_mode). Basically what needs to be done is -

In terminal -

    $ git rebase --interactive HEAD~23 <Enter> (23 is no. of your total commits on this PR. Increment this number if you commit more.)

This will open an editor where all your last 23 commits are listed.

Type `fixup` instead of `pick` written in front of each commit message. Save it.

Push forcefully -
    
    $ git push -f origin patch-10 <Enter> (patch-10 is your branch name)

**Comment by francescbassas** [comment]
Done. I fixup all the commits.

The wrong message displayed in relationship reports if only from date is selected happens also with the other select 2 widgets dates 'Start Date' and 'End Date'. So I think it would be better to treat it as a separated bug.

File: CRM/Report/Form/Contact/Relationship.php [hunk 1]
```diff
@@ -774,4 +783,48 @@ public function buildValidityQuery($valid) {
...
+
+    if ($from && $to) {
+      return CRM_Contact_BAO_Query::getRelationshipActivePeriodClauses($from, $to, FALSE);
+    }
+    elseif ($from) {
+      return CRM_Contact_BAO_Query::getRelationshipActivePeriodClauses($from, NULL, FALSE);
+    }
+    elseif ($to) {
+      return CRM_Contact_BAO_Query::getRelationshipActivePeriodClauses(NULL, $to, FALSE);
+    }
```
Comment (jitendrapurohit) [inline CRM/Report/Form/Contact/Relationship.php]: Can we remove the additional `elseif()`(line 821 - 826) here with `&&` replaced by `||` in line 818 ? I think `getRelationshipActivePeriodClauses()` should take care of proper return clause as the check for $from & $to is done there too.

File: CRM/Contact/BAO/Query.php [hunk 2]
```diff
@@ -4195,6 +4198,62 @@ public function addRelationshipDateClauses($grouping, &$where) {
...
+    elseif (!empty($dateValueLow)) {
+      $dateValueLowFormated = date('Ymd', strtotime($dateValueLow[2]));
+      $where[$grouping][] = self::getRelationshipActivePeriodClauses($dateValueLowFormated, NULL, TRUE);
+      $this->_qill[$grouping][] = (ts('Relationship was active after')) . " " . CRM_Utils_Date::customFormat($dateValueLowFormated);
+    }
+    elseif (!empty($dateValueHigh)) {
+      $dateValueHighFormated = date('Ymd', strtotime($dateValueHigh[2]));
+      $where[$grouping][] = self::getRelationshipActivePeriodClauses(NULL, $dateValueHighFormated, TRUE);
+      $this->_qill[$grouping][] = (ts('Relationship was active before')) . " " . CRM_Utils_Date::customFormat($dateValueHighFormated);
+    }
```
Comment (jitendrapurohit) [inline CRM/Contact/BAO/Query.php]: Can we remove `$where[$grouping][]` statement from all the three condition and place it after` if else if else if` (after line 4228) ?

File: CRM/Report/Form/Contact/Relationship.php (hunk 1 above)
Comment (francescbassas) [inline CRM/Report/Form/Contact/Relationship.php]: Done, see https://github.com/civicrm/civicrm-core/pull/10333/commits/02633eb09e38eb74dde0e6e0009e727317e76797

File: CRM/Contact/BAO/Query.php (hunk 2 above)
Comment (francescbassas) [inline CRM/Contact/BAO/Query.php]: Done, see https://github.com/civicrm/civicrm-core/pull/10333/commits/02633eb09e38eb74dde0e6e0009e727317e76797

**Review by jitendrapurohit — APPROVED** [review]
works fine after the final changes.

**Comment by francescbassas** [comment]
Thanks. I added some help to the advanced search form. Now are appearing failing checks but seems to be unrelated with my changes.

**Comment by francescbassas** [comment]
jenkins, test this please

**Comment by francescbassas** [comment]
@jitendrapurohit any plans to push it for 4.7.21?

**Comment by francescbassas** [comment]
@colemanw or @eileenmcnaughton it is possible to push this for 4.7.21?

**Comment by colemanw** [comment]
Code looks good and tested by Jitendra. Merging.

**Pull request merged**